CIPs/CIP-11/CIP-11.md

```diff
@@ -0,0 +1,261 @@
+
+By standardizing how data sets are associated to a DID, IDX solves discovery, routing, and resolution. This enables any app to lookup a DID and seamlessly interact with its resources regardless of where they exist or which application first created the data. As a result, IDX can unlock true identity-centric interoperability across the web.
+
+IDX achieves this by specifying data set **definitions** and **references**. A definition is created by a developer to describe a data set for their application. When a user arrives at the application a unique reference is created for the data set as defined in the definition. This means that each user will get a unique reference which they control access and privacy for, while only one global definition exists for any given data set.
+
+## Motivation
+
+IDX aims to:
+
+- Elevate DIDs into complete identities with the full set of resources and capabilities needed to be broadly useful
+- Provide a DID-centric framework for indexing data sets that is flexible, extensible, decentralized, cross-platform, and DID-agnostic
+- Simplify data set discovery and routing, wherever resources may exist (centralized servers, decentralized networks, etc)
+- Provide a standard interface for developers to use when interacting with any DID
+- Enable DIDs to easily integrate with other technologies, networks, platforms, and wallets
+- Make it easy to build with decentralized identities
+
+
+## Specification
+
+IDX is a Ceramic document which contains mappings from **definitions** to **references**. A *definition* is a Ceramic document created by a developer that describes a *data set*. This document contains a title and description, as well as a schema document that the *reference* has to conform to. In addition to this, the *definition* document may also include configuration options unique to the given data set. A *reference* is also a Ceramic document which has to use the schema defined in the *reference*. This document may contain data directly, but it could also contain pointers to data that lives elsewhere (e.g. in other Ceramic document, on other p2p databases, or hosted databases). Each user should only ever have one instance of any unique *definition*.
+
+The *definition-reference* pair can be stored publicly or encrypted. The former is desirable for data sets which require public discoverability. The latter is useful when possession of the data set should be kept secret. The encrypted *reference* should be encrypted using a unique symmetric key. This enables the data set to be selectively disclosed by asymmetrically encrypting the symmetric key to a given recipient.
+
+![IDX Example](./assets/idx-example.png)
+
+### IDX Document
+
+The IDX document simply consist of a map from strings to DocIDs. For public data sets the *key* in this map is the DocID (without the prepended `ceramic://`) of the *definition* document, and the *value* is the DocID (with `ceramic://`) of the *reference* document.
```

Comment:
I think we should enforce validation in the schema to avoid possibly invalid keys here.

The nice thing if both DocIDs and encrypted data keys are encoded in base 36 is that we can validate based on this.
We should also use a multicodec when setting an encrypted data key (could be simply `raw`), this way we can easily differentiate from DocIDs.

I think the logic would be something like:
1. Regexp matching base36 encoding byte followed by valid base 36 characters
1. Check second byte:
    - if matching DocID multicodec, it's not encrypted
    - if matching raw multicodec, follow encrypted logic
    - otherwise it's an invalid key per IDX spec

Reply:
I don't think it's that easy to validate base36 using regexp unfortunately. We could add a max length and enforce that only `[0-9][a-z]` is used however. Would want to see some examples from the DocID implementation before deciding on length though.

Reply:
It seems that `f` is the character used for base 16 so I imagine we can have a regexp like `/^fa|b[0-9a-z]+$/` (taking `a` and `b` as example codes for DocID or raw bytes) to validate the keys, no?

Reply:
`base36` is actually `k`. Might be possible to do this, but we need examples from the implementation first.

Reply:
As I'm going through the necessary changes in the IDX lib implementation I think we really need to have a way to validate these keys so that IDX can throw useful errors when invalid, otherwise it will be way harder to properly catch possible issues and implement the necessary logic for encrypted data keys.

Reply:
Not sure i follow why it would be difficult to catch errors? If DocID decoding fails, then check check if it's a base36 encoded string?

Reply:
Well we go from having one way to identify a document (the `ceramic://...` URL) to 3 (URL, string or encoded key) with possible associated logic and conversions all over the place, for example the Ceramic interactions need the URL form but the index keys should use the string form, so I think we need to at least have a good way to identify the inputs and outputs at the edges.

Reply:
So only `ceramic://` should be a valid "link" to another ceramic document. In the *key* of the map we can't really have links anyway and if you ever where to specify a path, e.g. `<idx-docid>/<def-docid>/<prop-in-reference>` it wouldn't really make sense to include `ceramic://`. 
The random key is not really a way to identify a definition, rather a way to find a definition that you've decrypted from the keychain.

This is nuanced though. Let's maybe find some time for a live sync?

Reply:
Yes I think we need to define the expected behavior from the developer side, what would the APIs look like in these different cases? For example:
- `idx.get('profile', aliceDID)`
- `idx.set(profileDefinitionId, { name: 'Bob' })`
- `idx.set('my-encrypted-data', { data: cleartext })`: can we do that? what is the expected behavior for the library? is the encryption based on the matching definition for 'my-encrypted-data'?
- `idx.set('profile',  { name: 'Bob' }, { encryptedFor: [aliceDID] })`: can we do that? what if 'profile' is already set but not encrypted?
- `idx.get('random app key')`: need to be able to clearly state if the key is a definition alias, definition ID or encrypted key